### PR TITLE
TabView visual updates

### PR DESCRIPTION
**Pull request by beervoley**
## Description

This PR updates TabView to make selected tabs more visible and adds some logic to work around tab adjacency issues.
## Motivation and Context

TabView mostly relied on shadows to draw the outlines of the Tabs but due to the proposed color changes, background of selected Tab is semi transparent. Therefore, we now need to draw the outline of the Tabs with a stroke. It brings some difficulties and this PR aims to solve them.

## How Has This Been Tested?

Manually.
## Screenshots (if appropriate):
Rest, hover ((over unselected tab),  overflow (with hover over next scroll button):
image
image
image

Rest and hover (over unselected tab):
image
image

### CODE REVIEW

**Comment by beervoley** [comment]
@chingucoding can you pleas have a look?:)
May be you can suggest an easier way to deal with the stroke introduction and adjacency issues:)

**Comment by chingucoding** [comment]
Sure @beervoley , I'll take a look at this.

**Comment by chingucoding** [comment]
The resulting visuals don't look right:

Screenshot of selected TabViewItem with shadow bleeding out

The shadow looks strange and the content is being cut off, I don't think that's supposed to happen is it?

**Comment by mdtauk** [comment]
> The resulting visuals don't look right:
> 
> Screenshot of selected TabViewItem with shadow bleeding out
> 
> The shadow looks strange and the content is being cut off, I don't think that's supposed to happen is it?

Looks like the shadow is showing through a semi-transparent background

File: dev/TabView/TabView.xaml
Comment (chingucoding) [inline dev/TabView/TabView.xaml:646]: Is this SVG from design?

File: dev/TabView/TabView.xaml
Comment (chingucoding) [inline dev/TabView/TabView.xaml:648]: Why do we need this now?

**Comment by chingucoding** [comment]
Hovered TVI's are also only flush on one side:

Image showing TabViewItem only being flush on one side, but not on the other side

File: dev/TabView/TabView.cpp
Comment (chingucoding) [inline dev/TabView/TabView.cpp:246]: I think a more robust approach is just handling on load on the TVI and ask the TabView what to instead of injecting a loaded event that get's revoked.

Comment (chingucoding) [reply dev/TabView/TabView.cpp:246]: Also how would this deal with virtualization? What if the TVI get's recycled?

File: dev/TabView/TabView.xaml
Comment (chingucoding) [inline dev/TabView/TabView.xaml:692]: I think this was needed since we otherwise would have a gap in layout. Are we sure that this won't be an issue anymore?

**Review by chingucoding — CHANGES_REQUESTED** [review]

File: dev/TabView/TabView.xaml
Comment (beervoley) [reply dev/TabView/TabView.xaml:646]: yup

File: dev/TabView/TabView.xaml
Comment (beervoley) [reply dev/TabView/TabView.xaml:648]: Otherwise, how would we do it ?:)
image
We need two pieces here: an arc that will draw that little "slope" and a "triangle" that will fill the space below the slope with the background color of the selected tab

File: dev/TabView/TabView.xaml
Comment (beervoley) [reply dev/TabView/TabView.xaml:692]: I'll check

File: dev/TabView/TabView.xaml
Comment (chingucoding) [reply dev/TabView/TabView.xaml:648]: The previous embedded SVG did that in one single shape, arc as the top left, then the left border and the bottom border and fill that.

Comment (beervoley) [reply dev/TabView/TabView.xaml:648]: yes, but we can't just fill this time because we need a stroke on the outside

**Comment by beervoley** [comment]
> Hovered TVI's are also only flush on one side:
> 
> Image showing TabViewItem only being flush on one side, but not on the other side

what do you mean by being flush?:)

File: dev/TabView/TabView.cpp
Comment (chingucoding) [reply dev/TabView/TabView.cpp:246]: OOP approach as in Object Oriented Programming? Don't think that that get's violated. Either way these components need to talk with each other.

Also does the TabView listen to recycling of elements? I wasn't able to find anything in this PR that would account for that, so if an element get's recycled TabView doesn't seem to be aware of that.

**Comment by chingucoding** [comment]
I mean this: 
Image showing TabViewItem only being flush on one side, but not on the other side

There is a gap on one of the sides.

File: dev/TabView/TabView.xaml
Comment (chingucoding) [reply dev/TabView/TabView.xaml:648]: Ah I see, that makes sense.

File: dev/TabView/TabView.cpp
Comment (beervoley) [reply dev/TabView/TabView.cpp:246]: I don't think TabView does. Please, correct me if I'm wrong, but underlying ListView should take a care of that, shouldn't it?:)

**Comment by beervoley** [comment]
> > The resulting visuals don't look right:
> > Screenshot of selected TabViewItem with shadow bleeding out
> > The shadow looks strange and the content is being cut off, I don't think that's supposed to happen is it?
> 
> Looks like the shadow is showing through a semi-transparent background

yup, that's probably what it is. Let me see how we can fix it :)

File: dev/TabView/TabView.cpp
Comment (chingucoding) [reply dev/TabView/TabView.cpp:246]: Yes ListView handles recycling, but currently you specify the bottom border on an element based on the position. However if it get's recycled onto a different position, it doesn't have to be adjacent to the selected element anymore. So you could end up where the visuals would be out of sync since the actual position in relation to the selected element.

Comment (beervoley) [reply dev/TabView/TabView.cpp:246]: I see. However, the bottom border (it's corner radius) is being dictated by certain Visual State. When the item will be recycled, will the Visual State that was set before not "nullify" (become unset)? :)

Comment (chingucoding) [reply dev/TabView/TabView.cpp:246]: So if they get nullified, you won't update again then right? Since TabView doesn't listen to element creation, how would it be aware of an item being shown next the selected index?

Comment (beervoley) [reply dev/TabView/TabView.cpp:246]: The bottom strokes will be update on every selected index change :)

Comment (chingucoding) [reply dev/TabView/TabView.cpp:246]: But if you scroll the selected index out of the realization rect, you have lost the bottom stroke value since TabView now thinks "Yeah we set the bottom stroke, we are done". If you then scroll to the selected index, TabView doesn't update the stroke anymore correctly. So you could end up with wrongly set stroke values.

Comment (beervoley) [reply dev/TabView/TabView.cpp:246]: I see what you mean, I'll look into it :)

File: dev/TabView/TabView.cpp
Comment (almedina-ms) [inline dev/TabView/TabView.cpp:750]: You could store the difference between indexes in a variable and query for the absolute value between them, then you can nest an if to verify whether the difference is positive or not, that way you can save one call to UpdateBottonStrokes
```
auto difference = itemIndex - selectedIndex;

if (abs(difference) <= 1)
{
    if (difference > 0)
    {
        SetTabViewItemAdjacentState(selectedIndex + 2, L"NotAdjacent");
    }
    else
    {
        SetTabViewItemAdjacentState(selectedIndex - 2, L"NotAdjacent");
    }

    UpdateBottomStrokes(SelectedIndex());
}
```

**Comment by beervoley** [comment]
@chingucoding I've updated the code : ) we decided to go an easier route and remove the arcs altogether
Could you please have a look once again?:)

**Comment by chingucoding** [comment]
> @chingucoding I've updated the code : ) we've decided to go an easier route and remove the arcs altogether
> Could you please have a look once again?:)

Why exactly where the arcs removed? Was all this listening to loading events really related to that?

**Comment by beervoley** [comment]
> > @chingucoding I've updated the code : ) we've decided to go an easier route and remove the arcs altogether
> > Could you please have a look once again?:)
> 
> Why exactly where the arcs removed? Was all this listening to loading events really related to that?

some of it to the arcs, some of it to the tab separators :)

**Comment by chingucoding** [comment]
So element recycling/virtualizing would still be an issue since we don't know if an adjacent element is still the selected one when moving the elements out of the realization rect?

**Comment by beervoley** [comment]
> So element recycling/virtualizing would still be an issue since we don't know if an adjacent element is still the selected one when moving the elements out of the realization rect?

I've added OnLoad to the TabViewItem to deal with this:)

**Comment by beervoley** [comment]
@chingucoding there some tests failing because of the margins, paddings, font size changes inside the Tab so previous tests expect different scenarios. Do you mind if I try to update them?

**Comment by chingucoding** [comment]
> @chingucoding there are some tests failing because of the margins, paddings, font size changes inside the Tab so previous tests expect different scenarios. Do you mind if I try to update them?

Sure, update them.

File: dev/TabView/TabView.xaml
Comment (chingucoding) [inline dev/TabView/TabView.xaml:622]: What issues exactly would this create? Still trying to wrap my head around the removal of this since it's one of the things making this control more fluent (looking at edge for example).

File: dev/TabView/TabViewListView.cpp
Comment (chingucoding) [inline dev/TabView/TabViewListView.cpp:22]: Not sure why you removed this

File: dev/TabView/TabView.xaml
Comment (chingucoding) [inline dev/TabView/TabView.xaml:50]: @StephenLPeters @beervoley So should I close #3776 then?
If so, @beervoley can you please add "Closes #3736" to the PR description?

File: dev/TabView/TabView.xaml
Comment (chingucoding) [inline dev/TabView/TabView.xaml:369]: I think if you remove the arcs, you don't need these columns anymore, everything lives inside column 1 now.

File: dev/TabView/TabView.xaml
Comment (beervoley) [reply dev/TabView/TabView.xaml:622]: First of all, it bring lots of code changes to make it properly work because now we have strokes everywhere and to make adjacent tabs (to the selected tab) smoothly connect with the arcs we need to give them corner radius only on one side.
    Second problem, this arc require negative margin from both sides to bring adjacent tabs right next to the selected one - so when you have selected tab as the very first or very last tab - it brings the footer and header closer meaning there's different spacing between unselected and selected tab when they are being 1st/last tab. It makes it almost impossible to connect header/footer with the tab in a continuous line :)

Third problem, the background of the arc as well as the selected tab itself is semi-transparent what makes anything behind it affect it. So when we have unselected tab hover background applied next to the selected one - it will be visible on the arc because it's technically "behind" the arc :)

File: dev/TabView/TabView.cpp
Comment (chingucoding) [inline dev/TabView/TabView.cpp:225]: We do this on SelectedIndex and SelectedIndex - 1 because the separator is always on the right of the TabViewItem, so the separators surrounding the selected item are those of it himself and that to the left of himself right?

nit: Short comment explaining why we call with these 4 values

**Review by chingucoding — COMMENTED** [review]
Is it just me or does the separator on the right look brighter than the one on the left?

Image with TabViewItem in dark theme being hovered showing the separators surrounding the item having different brightness

Same for light theme, the right one is more "intensive":

Image with TabViewItem in light theme being hovered showing the separators surrounding the item having different brightness

File: dev/TabView/TabView.xaml
Comment (beervoley) [reply dev/TabView/TabView.xaml:50]: @chingucoding you can check it in, I'll merge the master after :)

File: dev/TabView/TabView.xaml
Comment (beervoley) [reply dev/TabView/TabView.xaml:369]: yes, you're right, will do

File: dev/TabView/TabView.cpp
Comment (beervoley) [reply dev/TabView/TabView.cpp:225]: yes, you're correct :)
I'll add an explaining comment

File: dev/TabView/TabView.xaml
Comment (chingucoding) [reply dev/TabView/TabView.xaml:622]: > First of all, it bring lots of code changes to make it properly work because now we have strokes everywhere and to make adjacent tabs (to the selected tab) smoothly connect with the arcs we need to give them corner radius only on one side.

Don't think you would need to do that, after all we didn't have to do it now.

> Second problem, these arc require negative margin from both sides to bring adjacent tabs right next to the selected one - so when you have selected tab as the very first or very last tab - it brings the footer and header closer meaning there's different spacing between unselected and selected tab when they are being 1st/last tab. It makes it almost impossible to connect header/footer with the tab in a continuous line :)

Could add a know or margin for this I guess, but wouldn't be much fun.

> Third problem, the background of the arc as well as the selected tab itself is semi-transparent what makes anything behind it affect it. So when we have unselected tab hover background applied next to the selected one - it will be visible on the arc because it's technically "behind" the arc :)

Oh so the tabs are semi transparent now? Didn't know that, that makes sense then.

Guess they are gone for good (or for worse) then...

**Comment by beervoley** [comment]
> Is it just me or does the separator on the right look brighter than the one on the left?
> 
> Image with TabViewItem in dark theme being hovered showing the separators surrounding the item having different brightness
> 
> Same for light theme, the right one is more "intensive":
> 
> Image with TabViewItem in light theme being hovered showing the separators surrounding the item having different brightness

I think it's because left separator is not totally inside the "hover background" but right one is fully in it, so it looks like it's more "intensive":)

File: dev/TabView/TabView.xaml
Comment (chingucoding) [reply dev/TabView/TabView.xaml:50]: I can't check in anything so I'll defer to @StephenLPeters or @ranjeshj on this :)

File: dev/TabView/TabView.xaml
Comment (beervoley) [reply dev/TabView/TabView.xaml:622]: yup, because of semi transparency we have to do face lots of issues :)
it's probably possible to make it work, but it will require tons of time to figure out exactly how :)

**Comment by chingucoding** [comment]
> I think it's because left separator is not totally inside the "hover background" but right one is fully in it, so it looks like it's more "intensive":)

But this isn't intentional is it? After all this notion of separator belonging to a tab is purely in xaml, nothing the user expects to be here.

File: dev/TabView/TabView.xaml
Comment (chingucoding) [reply dev/TabView/TabView.xaml:622]: @stmoy @StephenLPeters Is this worth a tracking issue to reintroduce?

**Comment by gabrielconl** [comment]
There was a negative margin that made the tab cover both separators but it was removed.

**Comment by gabrielconl** [comment]
`TabViewHeaderPadding` is currently set to `8,8,4,0`, which causes the right bottom corner of the last tab to be slightly cut (compare left to right):
image
I see that the add button is supposed to be closer to the tabs in the new styles. Could this be done in a different way, to avoid cutting the corner visual?

**Comment by beervoley** [comment]
> `TabViewHeaderPadding` is currently set to `8,8,4,0`, which causes the right bottom corner of the last tab to be slightly cut (compare left to right):
> image
> I see that the add button is supposed to be closer to the tabs in the new styles. Could this be done in a different way, to avoid cutting the corner visual?

The tabs will be totally rectangular now : ) Add button won't cut any corners.

**Comment by beervoley** [comment]
> > I think it's because left separator is not totally inside the "hover background" but right one is fully in it, so it looks like it's more "intensive":)
> 
> But this isn't intentional is it? After all this notion of separator belonging to a tab is purely in xaml, nothing the user expects to be here.

I'll see what I can do :)

File: dev/TabView/TabView.xaml
Comment (ranjeshj) [reply dev/TabView/TabView.xaml:622]: @chingucoding Yes! I do agree that we want the bottom arcs. We just haven't figured out a good solution due to the transparency and this is a less than ideal compromise.

File: dev/TabView/TabView.xaml
Comment (ranjeshj) [reply dev/TabView/TabView.xaml:50]: I already merged in #3776, so we should be good as long as we merge with master cleanly.

**Pull request closed without merging**
